File: firebase/functions/lib/community/create_community.dart
import 'dart:async';

import 'package:firebase_admin_interop/firebase_admin_interop.dart';
import 'package:firebase_functions_interop/firebase_functions_interop.dart';
import '../on_call_function.dart';
import '../utils/infra/firestore_utils.dart';
import '../utils/utils.dart';
import 'package:data_models/cloud_functions/requests.dart';
import 'package:data_models/events/event.dart';
import 'package:data_models/community/community.dart';
import 'package:data_models/community/membership.dart';
import 'package:data_models/admin/partner_agreement.dart';
import 'package:data_models/utils/utils.dart';

class CreateCommunity extends OnCallMethod<CreateCommunityRequest> {
  CreateCommunity()
      : super(
          'createCommunity',
          (json) => CreateCommunityRequest.fromJson(json),
        );

  @override
  Future<Map<String, dynamic>> action(
    CreateCommunityRequest request,
    CallableContext context,
  ) async {
    orElseUnauthorized(context.authUid != null);

    final Community? requestedCommunity = request.community;
    if (requestedCommunity == null) {
      throw HttpsError(
        HttpsError.failedPrecondition,
        'Must provide community information.',
        null,
      );
    }

    Community community = requestedCommunity;

    if ((community.name?.trim() ?? '').isEmpty) {
      throw HttpsError(
        HttpsError.failedPrecondition,
        'Name is required.',
        null,
      );
    }

    final communityCollection = firestore.collection('/community/');
    final communityDocRef = communityCollection.document();

    final communitiesWithMatchingId = await communityCollection
        .where(
          Community.kFieldDisplayIds,
          arrayContains: communityDocRef.documentID,
        )
        .get();
    final alreadyUsedMessage =
        'The URL display name ${communityDocRef.documentID} is already taken.';
    if (communitiesWithMatchingId.isNotEmpty) {
      throw HttpsError(HttpsError.failedPrecondition, alreadyUsedMessage, null);
    }

    final userId = context.authUid;

    community = community.copyWith(
      id: communityDocRef.documentID,
      creatorId: userId,
      profileImageUrl: community.profileImageUrl ??
          'https://picsum.photos/seed/${communityDocRef.documentID}-profile/512',
      bannerImageUrl: community.bannerImageUrl ?? '',
      communitySettings: const CommunitySettings(),
      eventSettings: EventSettings.defaultSettings,
      // Ensure displayIds is not empty
      // If displayIds is empty, use the document ID as the first displayId
      displayIds: [
        community.displayIds.isNotEmpty
            ? community.displayIds[0]
            : communityDocRef.documentID,
    ],);

    final agreementChangedFields = [PartnerAgreement.kFieldCommunityId];
    DocumentReference agreementRef;
    PartnerAgreement agreementUpdated;

    if (request.agreementId != null) {
      agreementRef =
          firestore.document('partner-agreements/${request.agreementId}');
      final agreementDoc = await agreementRef.get();
      orElseUnauthorized(agreementDoc.exists);
      final agreement = PartnerAgreement.fromJson(
        firestoreUtils.fromFirestoreJson(agreementDoc.data.toMap()),
      );
      agreementUpdated =
          agreement.copyWith(communityId: communityDocRef.documentID);
    } else {
      agreementRef = firestore.collection('partner-agreements').document();
      agreementUpdated = PartnerAgreement(
        id: agreementRef.documentID,
        communityId: communityDocRef.documentID,
        allowPayments: true,
      );
      agreementChangedFields.addAll(
        [PartnerAgreement.kFieldId, PartnerAgreement.kFieldAllowPayments],
      );
    }

    await firestore.runTransaction((transaction) async {
      final doc = await transaction.get(communityDocRef);

      if (doc.exists) {
        throw HttpsError(
          HttpsError.failedPrecondition,
          alreadyUsedMessage,
          null,
        );
      }

      transaction.set(
        communityDocRef,
        DocumentData.fromMap(
          firestoreUtils.toFirestoreJson(community.toJson()),
        ),
      );

      transaction.set(
        firestore.document(
          'memberships/$userId/community-membership/${communityDocRef.documentID}',
        ),
        DocumentData.fromMap(
          firestoreUtils.toFirestoreJson(
            Membership(
              communityId: communityDocRef.documentID,
              userId: userId!,
              status: MembershipStatus.owner,
            ).toJson(),
          ),
        ),
      );

      transaction.set(
        agreementRef,
        DocumentData.fromMap(
          jsonSubset(
            agreementChangedFields,
            firestoreUtils.toFirestoreJson(agreementUpdated.toJson()),
          ),
        ),
        merge: true,
      );
    });

    return CreateCommunityResponse(communityId: communityDocRef.documentID)
        .toJson();
  }
}

File: firebase/functions/src/community/create_community.ts
import * as functions from 'firebase-functions'
import { HttpsError } from 'firebase-functions/lib/providers/https'
import { firestore, firestoreUtils } from '../utils/infra/firestore_utils'
import { OnCallMethod } from '../on_call_function'
import {
    CreateCommunityRequest,
    CreateCommunityResponse,
    Community,
    Membership,
    MembershipStatus,
    PartnerAgreement,
    CommunitySettings,
} from '../types'
import { orElseUnauthorized, jsonSubset } from '../utils/utils'

export class CreateCommunity extends OnCallMethod<CreateCommunityRequest> {
    constructor() {
        super('createCommunity', (json) => json as CreateCommunityRequest)
    }

    async action(
        request: CreateCommunityRequest,
        context: functions.https.CallableContext
    ): Promise<CreateCommunityResponse> {
        orElseUnauthorized(context.auth?.uid != null)

        const requestedCommunity = request.community
        if (!requestedCommunity) {
            throw new HttpsError('failed-precondition', 'Must provide community information.')
        }

        let community = { ...requestedCommunity }

        if (!community.name?.trim()) {
            throw new HttpsError('failed-precondition', 'Name is required.')
        }

        const communityCollection = firestore.collection('/community/')
        const communityDocRef = communityCollection.doc()

        if (community.displayIds && community.displayIds.length > 0) {
            console.log(`Checking if displayIds ${community.displayIds} are already in use.`)
            const communitiesWithMatchingId = await communityCollection
                .where('displayIds', 'array-contains', community.displayIds[0])
                .get()
            const alreadyUsedMessage = `The URL display name ${community.displayIds[0]} is already taken.`
            if (!communitiesWithMatchingId.empty) {
                throw new HttpsError('failed-precondition', alreadyUsedMessage)
            }
        }

        const userId = context.auth!.uid

        community = {
            ...community,
            id: communityDocRef.id,
            creatorId: userId,
            profileImageUrl:
                community.profileImageUrl ??
                `https://picsum.photos/seed/${communityDocRef.id}-profile/512`,
            bannerImageUrl: community.bannerImageUrl ?? '',
            communitySettings: community.communitySettings ?? ({} as CommunitySettings),
            displayIds: community.displayIds?.length
                ? [community.displayIds[0]]
                : [communityDocRef.id],
        }

        const agreementChangedFields = ['communityId']
        let agreementRef: FirebaseFirestore.DocumentReference
        let agreementUpdated: PartnerAgreement

        if (request.agreementId) {
            agreementRef = firestore.doc(`partner-agreements/${request.agreementId}`)
            const agreementDoc = await agreementRef.get()
            orElseUnauthorized(agreementDoc.exists)
            const agreement = agreementDoc.data() as PartnerAgreement
            agreementUpdated = { ...agreement, communityId: communityDocRef.id }
        } else {
            agreementRef = firestore.collection('partner-agreements').doc()
            agreementUpdated = {
                id: agreementRef.id,
                communityId: communityDocRef.id,
                allowPayments: true,
            }
            agreementChangedFields.push('id', 'allowPayments')
        }

        await firestore.runTransaction(async (transaction) => {
            const doc = await transaction.get(communityDocRef)
            if (doc.exists) {
                throw new HttpsError(
                    'failed-precondition',
                    `Community with ID ${communityDocRef.id} already exists.`
                )
            }

            transaction.set(
                communityDocRef,
                firestoreUtils.toFirestoreJson(community as unknown as Record<string, unknown>)
            )

            transaction.set(
                firestore.doc(`memberships/${userId}/community-membership/${communityDocRef.id}`),
                firestoreUtils.toFirestoreJson({
                    communityId: communityDocRef.id,
                    userId,
                    status: MembershipStatus.owner,
                })
            )

            transaction.set(
                agreementRef,
                jsonSubset(
                    agreementChangedFields,
                    firestoreUtils.toFirestoreJson(
                        agreementUpdated as unknown as Record<string, unknown>
                    )
                ),
                { merge: true }
            )
        })

        return { communityId: communityDocRef.id }
    }
}
